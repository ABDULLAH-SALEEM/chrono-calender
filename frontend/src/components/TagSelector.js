import React, { useState } from "react";
import {
  Box,
  Chip,
  TextField,
  Autocomplete,
  Typography,
  FormHelperText
} from "@mui/material";

const defaultTags = [
  { label: "Work", value: "work" },
  { label: "Personal", value: "personal" },
  { label: "University", value: "university" },
  { label: "Meeting", value: "meeting" },
  { label: "Appointment", value: "appointment" },
  { label: "Birthday", value: "birthday" },
  { label: "Holiday", value: "holiday" },
  { label: "Travel", value: "travel" }
];

const TagSelector = ({ value = [], onChange, error, label = "Tags" }) => {
  const [inputValue, setInputValue] = useState("");

  // Ensure value is always an array
  const safeValue = Array.isArray(value) ? value : [];

  const handleChange = (event, newValue) => {
    onChange(newValue || []);
  };

  const handleInputChange = (event, newInputValue) => {
    setInputValue(newInputValue);
  };

  return (
    <Box data-testid="tag-selector-root">
      <Autocomplete
        multiple
        freeSolo
        options={defaultTags}
        value={safeValue}
        onChange={handleChange}
        inputValue={inputValue}
        onInputChange={handleInputChange}
        getOptionLabel={(option) => {
          if (typeof option === "string") {
            return option;
          }
          return option.label || option;
        }}
        isOptionEqualToValue={(option, value) => {
          if (typeof option === "string" && typeof value === "string") {
            return option === value;
          }
          if (typeof option === "object" && typeof value === "object") {
            return option.value === value.value;
          }
          return false;
        }}
        renderTags={(value, getTagProps) =>
          value.map((option, index) => {
            const label = typeof option === "string" ? option : option.label;
            return (
              <Chip
                variant="outlined"
                label={label}
                {...getTagProps({ index })}
                key={index}
                onClick={(event) => {
                  // Prevent the default behavior and call onChange directly
                  event.stopPropagation();
                  const newValue = value.filter((_, i) => i !== index);
                  onChange(newValue);
                }}
              />
            );
          })
        }
        renderInput={(params) => (
          <TextField
            {...params}
            label={label}
            placeholder="Select or type tags..."
            error={!!error}
            fullWidth
          />
        )}
        renderOption={(props, option) => (
          <Box component="li" {...props}>
            <Typography>{option.label}</Typography>
          </Box>
        )}
      />
      {error && <FormHelperText error>{error}</FormHelperText>}
    </Box>
  );
};

export default TagSelector;
