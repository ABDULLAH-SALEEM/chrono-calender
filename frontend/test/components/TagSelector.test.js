import React from "react";
import { render, screen } from "@testing-library/react";
import TagSelector from "../../src/components/TagSelector";

describe("TagSelector", () => {
  it("renders without crashing", () => {
    render(<TagSelector value={[]} onChange={() => {}} />);
    expect(screen.getByTestId("tag-selector-root")).toBeInTheDocument();
  });

  it("matches snapshot", () => {
    const { asFragment } = render(
      <TagSelector value={[]} onChange={() => {}} />
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
