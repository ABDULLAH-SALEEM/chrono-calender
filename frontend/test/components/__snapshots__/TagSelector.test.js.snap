// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TagSelector matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="MuiBox-root css-0"
    data-testid="tag-selector-root"
  >
    <div
      class="MuiAutocomplete-root css-o5oj2-MuiAutocomplete-root"
    >
      <div
        class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-cmpglg-MuiFormControl-root-MuiTextField-root"
      >
        <label
          class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined css-19qnlrw-MuiFormLabel-root-MuiInputLabel-root"
          data-shrink="false"
          for="«r2»"
          id="«r2»-label"
        >
          Tags
        </label>
        <div
          class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl MuiAutocomplete-inputRoot css-1blp12k-MuiInputBase-root-MuiOutlinedInput-root"
        >
          <input
            aria-autocomplete="list"
            aria-expanded="false"
            aria-invalid="false"
            autocapitalize="none"
            autocomplete="off"
            class="MuiInputBase-input MuiOutlinedInput-input MuiAutocomplete-input MuiAutocomplete-inputFocused css-16wblaj-MuiInputBase-input-MuiOutlinedInput-input"
            id="«r2»"
            placeholder="Select or type tags..."
            role="combobox"
            spellcheck="false"
            type="text"
            value=""
          />
          <fieldset
            aria-hidden="true"
            class="MuiOutlinedInput-notchedOutline css-18p5xg2-MuiNotchedOutlined-root-MuiOutlinedInput-notchedOutline"
          >
            <legend
              class="css-1n64csd-MuiNotchedOutlined-root"
            >
              <span>
                Tags
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
